Add the rule that Peter requested, and add more literal support

CaspianA1 opened the pull request:
- Some code cleanup.
- New rule that reports when a transaction doesn't use the database.
- Removed some false positives for SQL injection. Your passed-in query string can now also be (in part) a function expression, arrow function, null literal, bools, enums, bigints, array literal expressions, regexps, conditional expressions, and some object literal expressions (not all are supported yet), and simple object property  access.
- Moved the scripts to their own directory, and made a script to update the eslint plugin in every known DBOS repo that uses it. This includes installing the new version, running the linter, and making a new branch + PR.
- Built a logging system for debugging during development.

CaspianA1 committed:
Updates:
- Some code cleanup.
- New rule that reports when a transaction doesn't use the database.
- Removed some false positives for SQL injection. Your passed-in query string can now also be (in part) a function expression, arrow function, null literal, bools, enums, bigints, array literal expressions, conditional expressions, and some object literal expressions (not all are supported yet).

CaspianA1 committed:
Make some code more succinct

CaspianA1 committed:
Add some more to the static analysis rule docs

CaspianA1 committed:
Move all the scripts to their own directory

CaspianA1 committed:
Make a script to upgrade a whole bunch of repos at once (to save time in the future)

CaspianA1 committed:
Now skipping making a PR if no changes were made during upgrading

qianl15 commented on dbos-rules.ts:
```diff
@@ -531,18 +542,62 @@ const isSqlInjection: ErrorChecker = (node, fnDecl, isLocal) => {
+  let foundDatabaseUsage = false;
+
+  fnDecl.getBody()!.forEachDescendant((descendant, traversalControl) => {
+    if (transactionContextSymbol === getSymbol(descendant)) {
```
Can you explain a bit more on how you detect database usage?

Does this only check the usage of `ctxt`? To detect whether a function is using the database, we want to make sure it uses `ctxt.client`

CaspianA1 replied:
Yes, this only checks for usage of  the passed-in transaction context. What the function does is that it checks if that symbol is present anywhere in the function body.

There's only one field behind 'ctxt', which is the client, so I'm not sure it would be really necessary to check for client field usage as well. And also, if you checked for just ctxt.client, then there would be a bunch of false positives for when transactions call helper functions that take just a context as a parameter.

CaspianA1 replied:
Does that make sense?

CaspianA1 replied:
Also my in-flight WiFi is limited so there might be slight communication gaps - sending this from an iPad; just 20 minutes WiFi per device so

qianl15 replied:
There are more fields behind ctxt: it contains all the properties from the base context (https://docs.dbos.dev/api-reference/contexts#properties).

However, I do agree that we don't want too many false positives.

CaspianA1 replied:
Got it that makes sense - didn't see that before
I'll change it to check for a `ctxt.client` or `client`, and see how many false positives arise over the demo apps

CaspianA1 replied:
Also, I have a question (unrelated to that):

So the definition for `PrismaClient` in `dbos-transact` is here:

https://github.com/dbos-inc/dbos-transact/blob/396d78eeeb51acf133fce10eeedf637e7989b916/src/user_database.ts#L178

But in the bank demo app backend, there's a different definition of `PrismaClient` which is used:

https://github.com/dbos-inc/dbos-demo-apps/blob/2f5cc06d43922667865db5bc3ff7e39b322f0d4b/bank/bank-backend/src/workflows/accountinfo.workflows.ts#L5

and

https://github.com/dbos-inc/dbos-demo-apps/blob/2f5cc06d43922667865db5bc3ff7e39b322f0d4b/bank/bank-backend/src/workflows/txnhistory.workflows.ts#L22

Should I use the official one from Prisma, or the interface from `dbos-transact`? Just wondering since my current mocked interface for it in the tests is using the official one.

The [docs](https://docs.dbos.dev/api-reference/contexts#transactioncontextt) say that using the official one should be used; then why is there the other interface for Prisma in `dbos-transact`?

qianl15 replied:
The tricky thing with PrismaClient is that it's auto-generated: https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/introduction
Unlike other ORMs which are installed directly, the PrismaClient is unique per application. Whenever you update your Prisma schema, you have to re-generate the client. Then DBOS will need to dynamically load the generated client ([here](https://github.com/dbos-inc/dbos-transact/blob/main/src/dbos-executor.ts#L230)), which is different per app.

Therefore, we need to provide a "mock" interface to capture common function interfaces such as `$transaction` and `$queryRawUnsafe` for us to use Prisma internally. Those two functions always appear in everyone's PrismaClient.

CaspianA1 committed:
Remove a newline

CaspianA1 committed:
Reformat a comment. Remove a snippet of a test.

CaspianA1 committed:
Minor formatting tweaks

qianl15 commented on dbos-rules.ts:
```diff
@@ -99,6 +95,8 @@ Also, some `bcrypt` functions generate random data and should only be called fro
   // The keys are the ids, and the values are the messages themselves
   return new Map([
     ["sqlInjection", "Possible SQL injection detected. The parameter to the query call site traces back to the nonliteral on line {{ lineNumber }}: '{{ theExpression }}'"],
+    ["transactionDoesntUseTheDatabase", "A transaction should always use the database; otherwise, it is just wasting time"],
```
I'd soften this a bit:
```suggestion
    ["transactionDoesntUseTheDatabase", "Unused TransactionContext arg, consider using a communicator or a normal function."],
```

CaspianA1 replied:
Got it; will add this in my next commit

CaspianA1 committed:
Fix the useless-transaction detection. Make an error message kinder. Remove `UserDatabase` (not passed into transactions). Also remove the arg index thing. Make some mocked interfaces more strongly typed.

CaspianA1 committed:
Make a type param correct

CaspianA1 committed:
Fix a range

CaspianA1 committed:
Remove a snippet of a test

CaspianA1 committed:
Use a triple-equals in one place

CaspianA1 committed:
Now eliminating some false positives for no-database-using transactions

CaspianA1 committed:
Extend a test a bit

CaspianA1 committed:
Add a TODO

CaspianA1 committed:
Condense two comments

CaspianA1 committed:
Revise some comments

CaspianA1 committed:
Change the repos upgrade script a bit

CaspianA1 committed:
Fix some possible false positives

CaspianA1 committed:
Change a repo name for 2 scripts

CaspianA1 committed:
Add the next template

qianl15 reviewed (approved):
Looks good!

CaspianA1 committed:
Finish most of the final stuff for this PR.
- There is now logging.
- The SQL injection error is now more elaborate.
- Now making a table of symbols that is used to quickly find where a node's is used elsewhere in the code. Very handy.
- Now supporting limited object fields for SQL injection.
- Many more types of literal values as LR-values are supported now.

CaspianA1 committed:
Make some tests more robust

CaspianA1 committed:
Make the transaction-doesnt-use-the-database rule more efficient (it doesnt rely on whole-function-body navigation anymore)
